### PR TITLE
Added support for Nuvoton NCT6793D chip

### PR DESCRIPTION
**Pull request by dernerv**
Added support for Nuvoton NCT6793D chip. 
Tested on Asus ASUS Z170 Pro Gaming.

### CODE REVIEW

**Comment by jtugsop** [comment]
Would love to see this pulled to provide full support for this chipset so Kaby Lake would be fully supported.

File: Hardware/LPC/NCT677X.cs [hunk 1]
```diff
@@ -414,7 +415,7 @@ public NCT677X(Chip chip, byte revision, ushort port, LPCPort lpcPort)
     public float?[] Controls { get { return controls; } }
 
     private void DisableIOSpaceLock() {
-      if (chip != Chip.NCT6791D)
+      if (chip != Chip.NCT6791D || chip != Chip.NCT6793D)
```
Comment (martynsmall) [inline Hardware/LPC/NCT677X.cs]: I believe this should be an `&&` not a `||`!

**Comment by martynsmall** [comment]
I tried making the same type of changes to add support for the NCT6795D chip (on an MSI Z270 PC Mate) using the ID `0xD352` and it seems to work well.  It picks up a few more temperature sensors than seem to be connected to anything as three of them never change but I guess this is down to what the chip actually supports.

Do you think you could possibly add this support into your PR as well so there aren't two very similar ones that will need merging in future?

**Comment by nokitakaze** [comment]
> approved

Ok. And what about always true condition in https://github.com/openhardwaremonitor/openhardwaremonitor/pull/884#discussion-diff-110516766R418 ?

**Comment by DarthDestroyer** [comment]
My approve doesnt do anything, but maybe it gets listen. 
And yes, it should be && instead of ||! 
But that is fixable, also when the pull gets accepted!

File: Hardware/LPC/NCT677X.cs (hunk 1 above)
Comment (jakobx13) [inline Hardware/LPC/NCT677X.cs]: @martynsmall are you sure?

Comment (martynsmall) [inline Hardware/LPC/NCT677X.cs]: 100%, as it stands the condition will always be true.  `chip` would have to be both `NCT6791D` and `NCT6793D` at the same time in order to fail the condition!

**Comment by myfreeweb** [comment]
I made a PR to the PR, for the 6795D https://github.com/dernerv/openhardwaremonitor/pull/1 — works on my MSI X370 SLI PLUS.

**Comment by TommyGKiwi** [comment]
Hi I read the Spec sheet for  NCT6796D which is in Asus RoG Hero Xi
I made some changes to at least reflect better the correct registers..... (see below)
(1) temperaturesSource, correct order versus alt temps reference 
(2)temperatureRegister were not referencing the correct locations for SMIOVT3-7
(3)PECI added to alternateTemperatureRegister



          temperatures = new float?[7];
          temperaturesSource = new byte[] {
            (byte)SourceNCT67XXD.PECI_0,
            (byte)SourceNCT67XXD.SYSTIN,
            (byte)SourceNCT67XXD.CPUTIN,
            (byte)SourceNCT67XXD.AUXTIN1,
            (byte)SourceNCT67XXD.AUXTIN2,
            (byte)SourceNCT67XXD.AUXTIN3,
            (byte)SourceNCT67XXD.AUXTIN4
          };

          temperatureRegister = new ushort[]
            { 0x027, 0x150, 0x670, 0x672, 0x674, 0x676, 0x678 };
          temperatureHalfRegister = new ushort[]
            { 0, 0x151, 0x671, 0x673, 0x675, 0x677, 0x679 };              
          temperatureHalfBit = new int[]
            { -1, 7, 7, 7, 7, 7, 7 };
          temperatureSourceRegister = new ushort[] 
            { 0x621, 0x622, 0xC26, 0xc27, 0xc28, 0xc29, 0xc2a };

          alternateTemperatureRegister = new ushort?[] 
            {0x720, 0x491, 0x490, 0x492, 0x493, 0x494, 0x495 };

File: Hardware/LPC/Chip.cs [hunk 2]
```diff
@@ -43,6 +43,8 @@ internal enum Chip : ushort {
     NCT610X = 0xC452,
     NCT6779D = 0xC560,
     NCT6791D = 0xC803,
```
Comment (Pavel-Sayekat) [inline Hardware/LPC/Chip.cs]: How you get that chip id from the CHIP model?

**Comment by TommyGKiwi** [comment]
I just related the C803 to the chip on the board, however the relationship
seems to be less clear.
The application seems to read bytes 20h & 21h for the information, but the
datasheets have other information
It seems it reads a copy of SuperIOs the same, I had Hero VIII and now a
Hero XI, The code (C803) works for both, but the chip ID is different (
NCT6791D & NCT6798D (but in website the latest is 6796D))
The 6796 datasheet seems to say the Chip ID is D4 and revision is 21

[image: image.png]
[image: image.png]

**Comment by TommyGKiwi** [comment]
Sorry Error, the chip ID are the same as the datasheets,  I was reading the
wrong master file chip.cs, 6798 seems to be D42B
so 6791 is C803

**Comment by moellerm** [comment]
This should be implemented in commit b4485c80d79476efcfc296b50125b9d7728560a5

**Pull request closed without merging**